include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -0,0 +1,661 @@
+                    const IndexSet              &column_parallel_partitioning,
+                    const SparsityPatternType   &sparsity_pattern,
+                    const bool                   exchange_data,
+                    const MPI_Comm               communicator,
+                    Teuchos::RCP<MapType<Node>> &column_space_map,
+                    Teuchos::RCP<MatrixType<Number, Node>> &matrix)
+      {
+        // release memory before reallocation
+        matrix.reset();
+
+        // Get the Tpetra::Maps
+        Teuchos::RCP<MapType<Node>> row_space_map =
+          row_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
+
+        column_space_map =
+          column_parallel_partitioning.make_tpetra_map_rcp(communicator, false);
+
+        if (column_space_map->getComm()->getRank() == 0)
+          {
+            AssertDimension(sparsity_pattern.n_rows(),
+                            row_parallel_partitioning.size());
+            AssertDimension(sparsity_pattern.n_cols(),
+                            column_parallel_partitioning.size());
+          }
+
+        // if we want to exchange data, build a usual Trilinos sparsity pattern
+        // and let that handle the exchange. otherwise, manually create a
+        // CrsGraph, which consumes considerably less memory because it can set
+        // correct number of indices right from the start
+        if (exchange_data)
+          {
+            SparsityPattern trilinos_sparsity;
+            trilinos_sparsity.reinit(row_parallel_partitioning,
+                                     column_parallel_partitioning,
+                                     sparsity_pattern,
+                                     communicator,
+                                     exchange_data);
+            matrix =
+              Utilities::Trilinos::internal::make_rcp<MatrixType<Number, Node>>(
+                trilinos_sparsity.trilinos_sparsity_pattern());
+
+            return;
+          }
+
+        IndexSet relevant_rows(sparsity_pattern.row_index_set());
+        // serial case
+        if (relevant_rows.size() == 0)
+          {
+            relevant_rows.set_size(row_space_map->getGlobalNumElements());
+            relevant_rows.add_range(0, row_space_map->getGlobalNumElements());
+          }
+        relevant_rows.compress();
+
+
+        const unsigned int n_rows = relevant_rows.n_elements();
+        std::vector<TrilinosWrappers::types::int_type> ghost_rows;
+        Teuchos::Array<size_t>                         n_entries_per_row(
+          row_space_map->getLocalNumElements());
+        for (unsigned int i = 0, own = 0; i < n_rows; ++i)
+          {
+            const TrilinosWrappers::types::int_type global_row =
+              relevant_rows.nth_index_in_set(i);
+            if (row_space_map->isNodeGlobalElement(global_row))
+              n_entries_per_row[own++] =
+                sparsity_pattern.row_length(global_row);
+          }
```

Comment:
This is not efficient, because the lookup `nth_index_in_set()` is expensive. I recognize that this is pre-existing, but take a look at #16398 to see how this can easily be addressed.

Reply:
I did not know that. I applied that patch.